### PR TITLE
kvserver: disable eager replicate enqueue on span cfg

### LINKED ISSUE #108724: kvserver: replicate enqueue on span config update is expensive
**Describe the problem**

In #100349, we began enqueuing replicas (into the replicate queue) upon receiving span config updates. The problem is, in clusters with a larger number of replicas per node, the overhead of enqueuing replicas is significant—and occurs regularly, every 10 minutes when the PTS changes.

**Expected behavior**

Replicas are enqueued into the replicate queue, when there is a span config change which would cause a replication/lease change. The overhead of this enqueuing is less noticeable on nodes with 100k+ leaseholders.

**Additional data / screenshots**

PTS record updated on span configs every 10 minutes, which causes a spike in CPU due to `ShouldPlanChange` called on enqueuing into the replicate queue.

image
image

**Environment:**
Affects master, release-23.1 and release-23.1.9-rc

Jira issue: CRDB-30613

### PR DESCRIPTION
**Pull request by kvoli**
Replicas were enqueued into the replicate queue, upon the store
receiving a span config update which could affect the replica. The
replicate queue shouldQueue is relatively more expensive than other
queues.
Introduce the cluster setting
kv.enqueue_in_replicate_queue_on_span_config_update.enabled, which when
set to true, enables queuing up replicas on span config updates; when
set to false, disables queuing replicas on span config updates.
By default, this settings is set to false.

Resolves: https://github.com/cockroachdb/cockroach/issues/108724

Release note (ops change): Introduce the
kv.enqueue_in_replicate_queue_on_span_config_update.enabled cluster
setting. When set to true, stores in the cluster will enqueue replicas
for replication changes, upon receiving config updates which could
affect the replica. This setting is off by default. Enabling this
setting speeds up how quickly config triggered replication changes
begin, but adds additional CPU overhead. The overhead scales with the
number of leaseholders.

### CODE REVIEW

**Comment by kvoli** [comment]
Index out of bounds on `TestProtectedTimestamps`. 

<details>
<summary>Failure</summary>


```
--- FAIL: TestProtectedTimestamps (1.13s)
panic: runtime error: index out of range [-1] [recovered]
	panic: runtime error: index out of range [-1] [recovered]
	panic: runtime error: index out of range [-1] [recovered]
	panic: runtime error: index out of range [-1] [recovered]
	panic: runtime error: index out of range [-1]

goroutine 7051 [running]:
testing.tRunner.func1.2({0x62d4880, 0xc00c03ed50})
	GOROOT/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1399 +0x39f
panic({0x62d4880, 0xc00c03ed50})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:133 +0x2f5
panic({0x62d4880, 0xc00c03ed50})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x0?, {0x7ecdd90, 0xc0000c0070})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a
panic({0x62d4880, 0xc00c03ed50})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc005426460, {0x7ecdd90, 0xc0000c0070})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:521 +0x1dd
panic({0x62d4880, 0xc00c03ed50})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Server(...)
	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:90
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestProtectedTimestamps.func5()
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_protectedts_test.go:133 +0x1e5
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestProtectedTimestamps.func6.1()
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_protectedts_test.go:139 +0x26
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1()
	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:75 +0x4a
github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration(0xc00278cd78?, 0xc00278cd50)
	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213 +0x142
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError(0xc00278ce78, 0xc000603380?)
	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:81 +0x9d
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x7ec9910, 0xc0098204e0}, 0xc0098204e0?, 0xc002b01500?)
	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x58
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x7ec9910, 0xc0098204e0}, 0xc0098204e0?)
	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:42 +0x4a
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestProtectedTimestamps.func6(0x6743788?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_protectedts_test.go:138 +0x52
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestProtectedTimestamps(0xc0098204e0)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_protectedts_test.go:170 +0x685
testing.tRunner(0xc0098204e0, 0x6a973f8)
	GOROOT/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1493 +0x35f
```

</details>

**Review by andrewbaptist — COMMENTED** [review]
:lgtm: 

> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-:-Nbpe4Yx2IOMCak4rWmx:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @DarrylWong, @erikgrinaker, @kvoli, and @rachitgsrivastava)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-NbpclhX0yDeWz6MtvYl:-NbpclhX0yDeWz6MtvYm:bkdu3ah) ([raw file](https://github.com/cockroachdb/cockroach/blob/703a9d54d80966c88babce09349d5c54ec6c700c/pkg/kv/kvserver/replicate_queue.go#L110)):*
> ```Go
> var EagerReplicateEnqueueOnSpanConfigUpdateEnabled = settings.RegisterBoolSetting(
> 	settings.SystemOnly,
> 	"kv.eager_replicate_enqueue_on_span_config_update.enabled",
> ```

nit: This name is a little hard to parse. Maybe change it (and the variable name) to:
`kv.enqueue_to_replicate_queue_on_span_config_update.enabled`
___
*[`pkg/kv/kvserver/store.go` line 2451 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-NbpdDJf5lVDGDowj6ty:-NbpdDJf5lVDGDowj6tz:b-8cy5pt) ([raw file](https://github.com/cockroachdb/cockroach/blob/703a9d54d80966c88babce09349d5c54ec6c700c/pkg/kv/kvserver/store.go#L2451)):*
> ```Go
> 			// (shouldQueue), because it scales with the number of stores, and
> 			// performs more checks.
> 			if EagerReplicateEnqueueOnSpanConfigUpdateEnabled.Get(&s.GetStoreConfig().Settings.SV) {
> ```

nit: Move the loading of `EagerReplicateEnqueueOnSpanConfigUpdateEnabled` outside the `VisitKeyRange` loop so it doesn't have to re-read the setting for every replica.

**Review by kvoli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108725)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist, @arulajmani, @DarrylWong, @erikgrinaker, and @rachitgsrivastava)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-NbpclhX0yDeWz6MtvYl:-NbpkBLB7y4881mVL5K9:bikp7d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/703a9d54d80966c88babce09349d5c54ec6c700c/pkg/kv/kvserver/replicate_queue.go#L110)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

nit: This name is a little hard to parse. Maybe change it (and the variable name) to:
`kv.enqueue_to_replicate_queue_on_span_config_update.enabled`
</blockquote></details>

Agreed, the naming is a bit off. Updated to your suggestion.
___
*[`pkg/kv/kvserver/store.go` line 2451 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-NbpdDJf5lVDGDowj6ty:-NbpkH2u8Nzr7_ub5EBK:buc25uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/703a9d54d80966c88babce09349d5c54ec6c700c/pkg/kv/kvserver/store.go#L2451)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

nit: Move the loading of `EagerReplicateEnqueueOnSpanConfigUpdateEnabled` outside the `VisitKeyRange` loop so it doesn't have to re-read the setting for every replica.
</blockquote></details>

Good idea, moved outside the visitor.

**Review by arulajmani — APPROVED** [review]
> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108725)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist, @DarrylWong, @erikgrinaker, @kvoli, and @rachitgsrivastava)

___
*[`-- commits` line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-Nbprd9j-wqSG9O8XpEv):*
nit: s/queue/queue's/g
___
*[`pkg/kv/kvserver/replicate_queue.go` line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-NbpclhX0yDeWz6MtvYl:-NbprwXA2zVEckAjxX1i:b-ttk5ks) ([raw file](https://github.com/cockroachdb/cockroach/blob/703a9d54d80966c88babce09349d5c54ec6c700c/pkg/kv/kvserver/replicate_queue.go#L110)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Agreed, the naming is a bit off. Updated to your suggestion.
</blockquote></details>

Should this be "enqueue_in ..." instead?
___
*[`pkg/kv/kvserver/client_protectedts_test.go` line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-NbptMzkEserV24Wgs3O:-NbptMzkEserV24Wgs3P:bt9vr7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/74db0915d9ad685c59de463f1046d7cf0eab4aff/pkg/kv/kvserver/client_protectedts_test.go#L132)):*
> ```Go
> 		startKey := getTableStartKey()
> 		// There's only one server, so there's no point searching for which server
> 		// the leaseholder is on, it could only be on s0.
> ```

Should we pull this out into its own commit/PR? That way, we can backport this to 22.2 and 22.1 as well.

**Review by kvoli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108725)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist, @arulajmani, @DarrylWong, @erikgrinaker, and @rachitgsrivastava)

___
*[`-- commits` line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-Nbprd9j-wqSG9O8XpEv):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/queue/queue's/g
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvserver/replicate_queue.go` line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-NbpclhX0yDeWz6MtvYl:-NbpwjGX9v4pT2gMBURL:b-3d2usz) ([raw file](https://github.com/cockroachdb/cockroach/blob/703a9d54d80966c88babce09349d5c54ec6c700c/pkg/kv/kvserver/replicate_queue.go#L110)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this be "enqueue_in ..." instead?
</blockquote></details>

`to` or `in` both seem correct to me, with `in` being more common.

Updated to `in`.
___
*[`pkg/kv/kvserver/client_protectedts_test.go` line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108725#-NbptMzkEserV24Wgs3O:-NbpwI0-2ZYwnOIvSfi_:bvukrbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/74db0915d9ad685c59de463f1046d7cf0eab4aff/pkg/kv/kvserver/client_protectedts_test.go#L132)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we pull this out into its own commit/PR? That way, we can backport this to 22.2 and 22.1 as well.
</blockquote></details>

This test fails without this code, even with the setting change I added after the cluster starts up and table is created. See the above comment mentioning the failure. 

I split this change out into an earlier commit.

**Comment by kvoli** [comment]
I'll wait for @erikgrinaker to take a look. Thanks for the reviews.

**Comment by knz** [comment]
Some questions I was expecting answers to in the release note:

- when the setting is disabled, when are changes to zone configs effected?

- what are the possible downsides?  (If there is no downside, why does the setting exist?)

**Review by erikgrinaker — APPROVED** [review]
LGTM, thanks. Should we default this to `true` on `master`, and work through/vet the performance issues?

I'll also join the bikeshed party: should we use the existing `kv.replicate_queue.` prefix here? Also, I recently discovered that the linter accepts both `_enabled` and `.enabled`, so `_enabled` might be better if this is a one-off setting rather than a group of settings. Take it or leave it.

**Comment by knz** [comment]
> so _enabled might be better if this is a one-off setting rather than a group of settings.

You don't know when a setting might become a group _in the future_. IMHO `.enabled` is the better choice by default.

**Comment by erikgrinaker** [comment]
> > so _enabled might be better if this is a one-off setting rather than a group of settings.
> 
> You don't know when a setting might become a group _in the future_. IMHO `.enabled` is the better choice by default.

That argument applies for any setting though, not just booleans. We generally don't create a new group for every setting.

**Comment by kvoli** [comment]
> * when the setting is disabled, when are changes to zone configs effected?

This is highly dependent on the number of ranges, store scanner env settings etc. There is no behavior change with the setting disabled, so I didn't include the default.

> what are the possible downsides? (If there is no downside, why does the setting exist?)

Updated the release note with the tradeoff the setting enables.

**Comment by kvoli** [comment]
> LGTM, thanks. Should we default this to `true` on `master`, and work through/vet the performance issues?

I'll open a follow up PR doing this.


> 
> I'll also join the bikeshed party: should we use the existing `kv.replicate_queue.` prefix here? Also, I recently discovered that the linter accepts both `_enabled` and `.enabled`, so `_enabled` might be better if this is a one-off setting rather than a group of settings. Take it or leave it.

I'm going to leave as is for now. I think the setting has a reasonable enough name, and I don't see us adding a boat load more replicate queue specific cluster settings.


TYFTRs!

**Comment by kvoli** [comment]
bors r=andrewbaptist,erikgrinaker,arulajmani
